### PR TITLE
KAFKA-15403: Refactor @Test(expected) annotation with assertThrows

### PR DESCRIPTION
**Pull request by Taher-Ghaleb**
I am working on research that investigates test smell refactoring in which we identify alternative implementations of test cases, study how commonly used these refactorings are, and assess how acceptable they are in practice.

In particular, this smell is when annotation, `@Test(expected)`, is used for exception handling instead of assertion, `assertThrows`.

We would like to get your feedback on this particular test refactoring. Thanks in advance for your input.

### CODE REVIEW

**Comment by jolshan** [comment]
Hey @Taher-Ghaleb -- thanks for contributing. I was curious the rationale for why the annotation should be replaced by assertThrows. I personally don't have a strong opinion, but was wondering more background on why this is a "smell."

**Comment by Taher-Ghaleb** [comment]
Thanks @jolshan for your response. `assertThrows` makes the verification of exceptions clearer and more intuitive, thus improving code readability compared to the annotation approach. [It is considered a test smell in the research literature](https://ieeexplore.ieee.org/abstract/document/9769994/). One possible research is due to developers not keeping up to date with recent versions of testing frameworks. 

Therefore, we aim in our research to get an idea of whether such a refactoring is acceptable in practice, and whether developers prefer to keep using `@Test(expected)` in certain testing scenarios. Thanks.

**Comment by vamossagar12** [comment]
Thanks for the explanation @Taher-Ghaleb . This was a nice thing to know! I have one nit level comment which I have added. BTW, any reason you chose this class? I can see a few other classes following this @Test(expected) pattern.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StoreToProcessorContextAdapterTest.java [hunk 1]
```diff
@@ -43,63 +45,87 @@ public void setUp() {
         context = StoreToProcessorContextAdapter.adapt(delegate);
     }
 
-    @Test(expected = UnsupportedOperationException.class)
+    @Test
     public void shouldThrowOnCurrentSystemTime() {
-        context.currentSystemTimeMs();
+        assertThrows(UnsupportedOperationException.class, () -> {
+            context.currentSystemTimeMs();
+        });
```
Comment (vamossagar12) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StoreToProcessorContextAdapterTest.java]: We don't need to include the statement in curly braces given this is a single line. Typically you will have IDEs throwing warnings like 
`Statement lambda can be replaced with expression lambda `
This comment would be applicable to all changed lines.

**Review by vamossagar12 — CHANGES_REQUESTED** [review]

**Comment by Taher-Ghaleb** [comment]
Thanks @vamossagar12 for your response.

Yes, there are other cases with such a pattern, but I aimed in this PR to get your impression about this kind of change first. I can do the same for other classes as well. Let me know please. Thanks.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StoreToProcessorContextAdapterTest.java (hunk 1 above)
Comment (Taher-Ghaleb) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/StoreToProcessorContextAdapterTest.java]: Thanks for pointing this out. I agree. I will fix this shortly.

Comment (vamossagar12) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/StoreToProcessorContextAdapterTest.java]: Thank you 👍

**Comment by vamossagar12** [comment]
@Taher-Ghaleb , doing it in an incremental manner makes sense and as such I am in favour of the change. Since you have modified tests within streams, if we could refactor all such usage patterns within Streams in this PR, that would be great. I could find only one more instance of this within Streams [here](https://github.com/apache/kafka/blob/ff3e6842ff99a600fc02e69ebefb09eef93decb3/streams/src/test/java/org/apache/kafka/streams/processor/internals/GlobalProcessorContextImplTest.java#L217), so that makes it easier. Then we can have streams committers review/merge this. Let me know what you think.

**Comment by Taher-Ghaleb** [comment]
Sure, sounds good. I will do it now. Please keep me posted. Thanks.

**Comment by vamossagar12** [comment]
Test Failures seem unrelated. cc @cadonna can you plz take a look at this PR. Thanks!

**Review by vamossagar12 — APPROVED** [review]
LGTM

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StoreToProcessorContextAdapterTest.java [hunk 2]
```diff
@@ -26,10 +26,12 @@
 import org.easymock.MockType;
 import org.junit.Before;
 import org.junit.Test;
+import static org.junit.Assert.assertThrows;
 import org.junit.runner.RunWith;
 
 import java.time.Duration;
 
+
```
Comment (jolshan) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StoreToProcessorContextAdapterTest.java]: nit: can we remove this extra space here?

**Comment by jolshan** [comment]
@Taher-Ghaleb can you pull in the latest changes from master and resolve conflicts?

**Comment by Taher-Ghaleb** [comment]
Hi @jolshan. I resolved the conflict and removed the extra space. Thanks.

**Comment by jolshan** [comment]
Thanks. Test failures are unrelated. I will go ahead and merge.

**Comment by Taher-Ghaleb** [comment]
Thanks, @jolshan.

**Pull request merged**